File: src/DataScience/MATLAB/presentation_abstraction_control.m
function result = presentation_abstraction_control
%PRESENTATION_ABSTRACTION_CONTROL Coordinate hierarchical agents through controllers.
child = makeAgent("child", 40);
child = control(child, "increment", 2);
root = makeAgent("root", child.abstraction.value);
result = struct("childView", present(child), "rootView", present(root));
end

function agent = makeAgent(name, value)
agent = struct("name", name, "abstraction", struct("value", value));
end

function agent = control(agent, action, amount)
if action == "increment"
    agent.abstraction.value = agent.abstraction.value + amount;
end
end

function view = present(agent)
view = agent.name + ":view=" + string(agent.abstraction.value);
end

File: src/Enterprise/C#/patterns/PresentationAbstractionControl.cs
namespace Genkidama.PatternExamples;
public static class PresentationAbstractionControlExample { public static bool Run(){string View(string n,int v)=>$"{n}:view={v}";return View("child",42)=="child:view=42"&&View("root",42)=="root:view=42";} }
